File: 03_Arrays/Easy/C++/09_missing_number.cpp
#include <bits/stdc++.h>
using namespace std;

int my_partition(vector<int>& arr, int low, int high){
  int pivot = arr[low];
  int i = low;
  int j = high;

  while( i < j ){
    while( arr[i] <= pivot && i <= high - 1){
      ++i;
    }

    while( arr[j] > pivot && j >= low + 1){
      --j;
    }

    if(i < j){
      swap(arr[i], arr[j]);
    }
  }

  swap(arr[low], arr[j]);

  return j;
}

void qS(vector<int>& arr, int low, int high){
  if (low < high){
    int partition_index = my_partition(arr, low, high);
    qS(arr, low, partition_index - 1);
    qS(arr, partition_index + 1, high);
  }
}

void quickSort(vector<int>& arr){
  qS(arr, 0, arr.size()-1);
}

int missingNumber_brute(vector<int>& a, int N){
  
  for(int i=1; i<=N; ++i){ // range of values

    // initially assume, the num 'i' is missing
    bool isMissing = true;
    for(int j=0; j<a.size(); ++j){ // ptr to the values of arr

      if(a[j] == i){
        // if number is found in the array
        // set the missing flag to false
        isMissing = false;
        break;
      }
    }

    if(isMissing){
      return i;
    }
  }

  return N;
}

int missingNumber_myBrute(vector<int>& a, int N) {
    // a[]: [1 -> n]
    // sort
    // diff b/w consectuive elements would be 1
    // the moment it is greater, 
    // break the loop, and return smaller + 1
    quickSort(a); // TC -> (NlogN)

    // TC -> O(N)
    for(int i=1; i<a.size(); ++i){
        if(a[i]-a[i-1] > 1){
            return a[i-1] + 1;
        }
    }

    // Total TC -> O(NlogN + N) 
    return N;
}

// better
int missingNumber_better(vector<int>& a, int N){
// hash map
  vector<bool> visited(N, false);

  // visited index -> [0, N-1]
  //    with value -> {0}

  // if value of arr acts as index for the visited
  // then value with 0 at the end will be the our answer

  for(int i=0; i<a.size(); ++i){
    visited[a[i]] = true;
  }

  for(int i=1; i<=a.size(); ++i){
    if (!visited[i]){
      return i;
    }
  }

  return N;
}

// optimal: method: #1 (using sum of N natural nums)
int missingNumber_optimal_sumN(vector<int>& a, int N){

  long sumN = N*(N+1)/2;

  long sum = 0;
  int rem = sumN;
  for(int i=0; i<a.size(); ++i){
    rem = rem - a[i];
    sum += a[i];
  }

  return rem;
}

// optimal: method: #2 (using XOR)
int missingNumber_optimal_xor(vector<int>& a, int N){
  int xor_N = 0;
  int xor_a = 0;

  // a ^ a = 0
  // 0 ^ a = a
  // xor_a ^ xor_N -> (1^1)^(2^2)^...^(missing Num)^...^(N^N)
            //     -> 0 ^ (missing Num) ^ 0
            //     -> missing Num
  for(int i=0; i<a.size(); ++i){
    // xor_N = 1^2^...^a.size() (remeber: a.size() -> N-1)
    xor_N = xor_N ^ (i+1);

    // xor of elements in the array
    xor_a = xor_a ^ a[i];
  }

  // to compensate the remaining N in the xor
  xor_N = xor_N ^ N;
  return xor_a ^ xor_N;

}

int main(int argc, char* argv[]){
vector<int> v;

  int size;
  cin >> size;

  // enter sorted array
  for(int i=0; i<size; ++i){
    int a;
    cin >> a;
    v.push_back(a);
  }

  int target;
  cin >> target;

  // int missingNum = missingNumber_brute(v, target);
  // int missingNum = missingNumber_myBrute(v, target);
  // int missingNum = missingNumber_better(v, target);
  // int missingNum = missingNumber_optimal_sumN(v, target);
  int missingNum = missingNumber_optimal_xor(v, target);

  cout << "Missing Number is: " << missingNum;
return 0;
}

File: 03_Arrays/Easy/JavaScript/09_missingNumber.js
/**
 * Problem Statement: Given an array nums containing n distinct numbers  in the range [0, n],
 *                    return the only number in the range
 *                    that is missing from the array.
 */

/**
 * @param {number[]} nums
 * @return {number}
 */
const missingNumberBrute = function (nums) {
  const n = nums.length;

  for (let i = 0; i <= n; ++i) {
    // to pick the number in the range [0,1]
    let targetFound = false;
    // Linear Search
    for (let j = 0; j < n; ++j) {
      if (nums[j] === i) {
        targetFound = true;
        break;
      }
    }

    if (!targetFound) return i;
  }

  /**
   * TC -> O(N^2)
   * SC -> O(1)
   */
};

const missingNumberBetter = function (nums) {
  const n = nums.length;

  let hashMap = {}; // hashMap to store the elements

  for (let i = 0; i < n; ++i) {
    hashMap[nums[i]] = true;
  }

  for (let i = 0; i <= n; ++i) {
    // if i not found in the hashMap,
    // it is the missing number
    if (!hashMap[i] ?? true) {
      return i;
    }
  }

  /**
   * TC -> O(2N) ~ O(N)
   * Sc -> O(N)
   */
};

const missingNumberBetter_usingSort = function (nums) {
  const n = nums.length;
  let temp = [...nums];

  temp.sort((a, b) => a - b);
  let missingNum = -1;

  for (let i = 0; i < n; ++i) {
    if (temp[i] === i) continue;

    missingNum = i;
    break;
  }

  if (missingNum === -1) return n;
  return missingNum;
  /**
   * TC -> O(NlogN + N) ~ O(NlogN)
   * Sc -> O(N)
   */
};

const missingNumberOptimal_sum = function (nums) {
  let arraySum = 0;
  let actualSum = 0;
  for (let i = 0; i < nums.length; ++i) {
    arraySum += nums[i];
    actualSum += i;
  }
  
  actualSum += nums.length;

  return actualSum - arraySum;
};

const missingNumberOptimal_xor = function (nums) {
  let xor_arr = 0;
  let xor_num = 0;

  for(let i=0; i<nums.length; ++i) {
    xor_arr = xor_arr ^ nums[i];
    xor_num = xor_num ^ i;
  }

  xor_num = xor_num ^ nums.length;

  return xor_num ^ xor_arr;
}

let arr = [
  45, 35, 38, 13, 12, 23, 48, 15, 44, 21, 43, 26, 6, 37, 1, 19, 22, 3, 11, 32,
  4, 16, 28, 49, 29, 36, 33, 8, 9, 39, 46, 17, 41, 7, 2, 5, 27, 20, 40, 34, 30,
  25, 47, 0, 31, 42, 24, 10, 14,
];

console.log("Missing number(brute): ", missingNumberBrute(arr));
console.log("Missing number(better): ", missingNumberBetter(arr));
console.log(
  "Missing number(better using sort): ",
  missingNumberBetter_usingSort(arr)
);
console.log("Missing number(optimal sum): ", missingNumberOptimal_sum(arr));
console.log("Missing number(optimal xor): ", missingNumberOptimal_xor(arr));
